app/src/features/FhirResourceTree/resourceTreeUtils.ts

```diff
@@ -199,51 +262,174 @@ export const getParent = (
  */
 export const buildTree = (
+  nodeDefinition: DefinitionNode,
+  parentElementNode?: ElementNode
+): ElementNode => {
+  const currentNode = createElementNode(nodeDefinition, {
+    parentPath:
+      // If the parent is of choice kind (ie value[x]), we need to remove value[x] from its path to rightly set the child path instead
+      parentElementNode?.kind !== "choice"
+        ? parentElementNode?.path
+        : parentElementNode?.path.split(".").slice(0, -1).join("."),
+  });
+  const isCurrentNodeRoot = !currentNode.path.includes(".");
+
+  // If current nodeDefinition has slices, we add them to its children
+  nodeDefinition.sliceDefinitions.forEach((sliceDefinition, index) => {
+    if (
+      sliceDefinition.definition.max &&
+      (sliceDefinition.definition.max === "*" ||
+        +sliceDefinition.definition.max > 0)
+    ) {
+      const sliceElementNode = createElementNode(sliceDefinition, {
+        index: currentNode.isArray ? index : undefined,
+        parentPath: currentNode.path,
+      });
+      currentNode.children.push(sliceElementNode);
+      sliceDefinition.childrenDefinitions.forEach((sliceChildDefinition) => {
+        buildTree(sliceChildDefinition, sliceElementNode);
+      });
+    }
+  });
+
+  // If current node is array, has no children and has definitions, we add an item
+  if (
+    !isCurrentNodeRoot &&
+    currentNode.isArray &&
+    currentNode.children.length === 0 &&
+    nodeDefinition.childrenDefinitions.length > 0
+  ) {
+    const itemNode = createElementNode(nodeDefinition, {
+      index: currentNode.children.length,
+      parentPath: currentNode.path,
+    });
+    currentNode.children.push(itemNode);
+    nodeDefinition.childrenDefinitions.forEach((childNodeDefinition) => {
+      buildTree(childNodeDefinition, itemNode);
+    });
+  }
+
+  if (parentElementNode) {
+    const isParentRoot = !parentElementNode.path.includes(".");
+
+    // If parent is not of type array, we add its childrenDefinitions as children
+    if (!parentElementNode.isArray || isParentRoot) {
+      parentElementNode.children.push(currentNode);
+      nodeDefinition.childrenDefinitions.forEach((childDefinition) => {
+        buildTree(childDefinition, currentNode);
+      });
+    }
+  } else {
+    nodeDefinition.childrenDefinitions.forEach((childDefinition) =>
+      buildTree(childDefinition, currentNode)
+    );
+  }
+
+  return currentNode;
+};
+
+/**
+ * Mutates the `rootNodeDefinition` parameter to set the whole DefinitionNode tree
+ * @param elementDefinitions Snapshot's elementDefinition array
+ * @param rootNodeDefinition The DefinitionNode root of the tree
+ * @param previousElementNodeDefinition
+ */
+export const buildTreeDefinition = (
   elementDefinitions: IElementDefinition[],
-  rootNode: ElementNode,
-  previousElementNode: ElementNode
+  rootNodeDefinition: DefinitionNode,
+  previousElementNodeDefinition: DefinitionNode
```

Comment:
again, naming is confusing: I would remove "Element" --> `previousNodeDefinition`